crates/ide_assists/src/handlers/convert_into_to_from.rs

```diff
@@ -0,0 +1,295 @@
+use std::{iter, ops::RangeInclusive};
+
+use ide_db::{helpers::FamousDefs, traits::resolve_target_trait};
+use syntax::{
+    ast::{self, edit::AstNodeEdit, AstNode, NameOwner},
+    SyntaxElement,
+};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: convert_into_to_from
+//
+// Converts an Into impl to an equivalent From impl.
+//
+// ```
+// # //- /lib.rs crate:core
+// # pub mod convert { pub trait Into<T> { pub fn into(self) -> T; } }
+// # //- /lib.rs crate:main deps:core
+// # use core::convert::Into;
+// impl $0Into<Thing> for String {
+//     fn into(self) -> Thing {
+//         Thing {
+//             b: self.len(),
+//             a: self
+//         }
+//     }
+// }
+// ```
+// ->
+// ```
+// # use core::convert::Into;
+// impl From<String> for Thing {
+//     fn from(val: String) -> Self {
+//         Thing {
+//             b: val.len(),
+//             a: val
+//         }
+//     }
+// }
+// ```
+pub(crate) fn convert_into_to_from(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let impl_ = ctx.find_node_at_offset::<ast::Impl>()?;
+    let src_type = impl_.self_ty()?;
+    let ast_trait = impl_.trait_()?;
+
+    impl_is_an_into(&impl_, &ctx)?;
+
+    let dest_type = match &ast_trait {
+        ast::Type::PathType(path) => {
+            path.path()?.segment()?.generic_arg_list()?.generic_args().next()?
```

Comment:
Seeing the segment calls here and in some other places makes me wonder, does this code cover qualified types for both self type and trait arg type? As in does it work for `impl Into<some_mod::Foo> for some_mod::Bar {...}`?

Reply:
Good spot - added some tests to check this.   

This particular instance seems to be fine, but the other one was doing the wrong thing when faced with `for some_mod::Bar<SomeGenericParams>` - was swapping in `Bar` instead of `some_mod::Bar` when it found `Self`.  I've updated it to use `Module::find_use_path` instead of just trying to figure it out from the AST.